cedar-policy-validator/src/schema_file_format.rs

```diff
@@ -133,6 +143,193 @@ impl SchemaFragment {
     }
 }
 
+/// A common type reference resolver
+#[derive(Debug)]
+pub(crate) struct CommonTypeResolver<'a> {
+    /// The `SchemaFragment` to resolve
+    schema: &'a mut SchemaFragment,
+    /// The dependency graph among common type names
+    /// A common type name is prefixed with the namespace id where it's declared
+    graph: HashMap<Name, HashSet<Name>>,
+}
+
+impl<'a> CommonTypeResolver<'a> {
+    /// Construct a the resolver
+    pub(crate) fn new(schema: &'a mut SchemaFragment) -> Self {
+        let mut graph = HashMap::new();
+        for (ns, def) in schema.0.iter() {
+            for (ty_name, ty) in def.common_types.iter() {
+                let name = Name::from(ty_name.clone()).prefix_namespace_if_unqualified(ns.clone());
+                graph.insert(name, HashSet::from_iter(ty.common_type_references()));
+            }
+        }
+        Self { schema, graph }
+    }
+
+    /// Perform topological sort on the dependency graph
+    /// Let A -> B denote the RHS of type `A` refers to type `B` (i.e., `A`
+    /// depends on `B`)
+    /// topo_sort(A -> B -> C) produces [C, B, A]
+    /// If there is a cycle, a type name involving in this cycle is the error
+    /// It implements a variant of Kahn's algorithm
+    fn topo_sort(&self) -> std::result::Result<Vec<Name>, Name> {
```

Comment:
This is a nontrivial graph algorithm, thoughts on using a dependency like `petgraph` (76 million downloads on crates) instead of hand-rolling?  I'm mostly just worried that if there are subtle bugs in this algorithm, code review / tests / fuzzing might not catch them.

Reply:
I tried to use this crate and eventually gave up on it. The reason is that it requires the node type of its hash graphs to implement `Copy`, which is kind of hard for Cedar's `Name`.

Reply:
I also argue this algorithm is fairly standard. @andrewmwells-amazon what do you think?

Reply:
Ok, I'm fine with this.  We did hand-roll transitive closure rather than using an existing graph library, so there's precedent.